### PR TITLE
✨ Dummy data generator

### PR DESCRIPTION
**Pull request by parimalak**

### CODE REVIEW

File: dataservice/util/data_gen/data_generator.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
+from datetime import datetime
+from dateutil import relativedelta
+import uuid
+import random
+
+from dataservice.util.data_gen.data import *
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: `*` should be avoided.
You can look up python start imports for all the reasons it's bad.

File: dataservice/util/data_gen/data.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+the entities
+"""
+import random
+import csv
+
+NUM_PARTICIPANTS = 10
+
+# NUM_DEMOGRAPHICS = 1
+
+NUM_DIAGNOSES = random.randint(0, 35)
```
Comment (dankolbman) [inline dataservice/util/data_gen/data.py]: Setting these to a random number on import seems counter intuitive.
This way, all samples will have the same number of aliquots, but that number will change between runs.
Using a random number of children for each parent may be a better idea. Perhaps this number could be `MAX_DIAGNOSES` then

File: dataservice/util/data_gen/data_generator.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+from dataservice.api.participant.models import Participant
+from dataservice.api.demographic.models import Demographic
+from dataservice.api.diagnosis.models import Diagnosis
+from dataservice.api.sample.models import Sample
+from dataservice.api.aliquot.models import Aliquot
+from dataservice.api.sequencing_experiment.models import SequencingExperiment
+from dataservice.api.genomic_file.models import GenomicFile
+
+
+class DataGenerator(object):
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py:19]: Since you have a class to represent the generator's state, why no use class attributes instead of global variables?

File: dataservice/util/data_gen/data.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+NUM_SAMPLES = random.randint(0, 65)
+
+NUM_ALIQUOTS = random.randint(0, 65)
+
+NUM_SEQ_EXP = 1
+
+NUM_GENOMIC_FILES = 5
+
+# Demographics data
+race_list = [
```
Comment (dankolbman) [inline dataservice/util/data_gen/data.py]: Might be nice to put these in a `txt` file to stay consistent and have less clutter here.

**Review by dankolbman — CHANGES_REQUESTED** [review]
Looks like a lot of files are duplicated between `tests/` and `dataservice/utils/`?

File: dataservice/util/data_gen/data_generator.py (hunk 1 above)
Comment (znatty22) [reply dataservice/util/data_gen/data_generator.py]: Ditto

File: dataservice/util/data_gen/data_generator.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
+from datetime import datetime
+from dateutil import relativedelta
+import uuid
+import random
+
+from dataservice.util.data_gen.data import *
+from dataservice import create_app
+from dataservice.extensions import db
+from dataservice.api.participant.models import Participant
+from dataservice.api.participant.models import Participant
```
Comment (znatty22) [inline dataservice/util/data_gen/data_generator.py]: Remove duplicate import

File: dataservice/util/data_gen/data_generator.py (hunk 3 above)
Comment (znatty22) [reply dataservice/util/data_gen/data_generator.py:19]: Agreed, I think you could have class attributes to represent the default totals for each entity and then use them as default args for each create entity method. 

And then probably the rest of the data/inputs driving the generator should all be outside of the class. I see that this class has some inputs provided from data.py and other inputs hardcoded in. I know it would be tedious to move all the inputs out 😑 ... but then the generator is totally agnostic of the inputs its using to build the data. And then its also nice bc you could have different data input files for the generator. Maybe something for the next version though?

File: tests/dummy_data_generator/test_data_generator.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
+from datetime import datetime
```
Comment (znatty22) [inline tests/dummy_data_generator/test_data_generator.py]: I think you meant to remove the tests/dummy_data_generator folder right?

File: dataservice/util/data_gen/data.py (hunk 2 above)
Comment (znatty22) [reply dataservice/util/data_gen/data.py]: Agreed. These numbers should either be constant totals for each entity type or maximums so that you can randomize the total (bounded by max) of that entity at runtime not during import

**Review by znatty22 — COMMENTED** [review]
Looks pretty good! See comments on files

File: tests/dummy_data_generator/test_data_generator.py (hunk 6 above)
Comment (znatty22) [reply tests/dummy_data_generator/test_data_generator.py]: Remember to remove tests/dummy_data_generator

File: tests/dummy_data_generator/test_data_generator.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        for line in reader:
+            diagnosis_list.append(line[0])
+        progression_or_recurrence_list = ['yes', 'no', 'unknown']
+        diag_list = []
+        for i in range(total):
+
+            data = {
+                'external_id': 'diagnosis_{}'.format(i),
+                'diagnosis': random.choice(diagnosis_list),
+                'progression_or_recurrence':
```
Comment (znatty22) [inline tests/dummy_data_generator/test_data_generator.py]: Remove progression_or_recurrence. This field is no longer in the Diagnosis model

**Comment by znatty22** [comment]
I think in the future we may want to decouple the data input generation (creating/reading in choices, setting default maximums, etc) from the data generation/population (creating db objects and publishing to db). But this is probably fine for now in terms of an initial MVP

File: dataservice/util/data_gen/data_generator.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+        """
+        creates diagnoses
+        """
+        self.diag_list = []
+        for i in range(total):
+
+            self.data = {
+                'external_id': 'diagnosis_{}'.format(i),
+                'diagnosis': random.choice(self.diagnosis_list),
+                'progression_or_recurrence':
```
Comment (znatty22) [inline dataservice/util/data_gen/data_generator.py]: Remove progression_or_recurrence. This field is no longer in the Diagnosis model

File: tests/genomic_file/test_geonmic_file_models.py [hunk 9]
```diff
@@ -22,6 +22,7 @@ def test_create_and_find(self):
         """
```
Comment (dankolbman) [inline tests/genomic_file/test_geonmic_file_models.py:22]: Looks like a typo in this file's name

File: dataservice/util/data_gen/data_generator.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+    def _create_experiments(self, total):
+        """
+        Creates sequencing experiments
+        """
+
+        self.e_list = []
+        dt = datetime.now()
+        for i in range(total):
+
+            self.e_data = {
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: Could do with just assigning to `e_data` rather than a class property. I'm guessing you never actually need `self.e_data`?

File: dataservice/util/data_gen/data_generator.py [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+                'tissue_type': random.choice(self.tissue_type_list),
+                'composition': random.choice(self.composition_list),
+                'anatomical_site': random.choice(self.anatomical_site),
+                'age_at_event_days': random.randint(0, 32872),
+                'tumor_descriptor': random.choice(self.tumor_descriptor_list)
+            }
+            self.s_list.append(
+                Sample(
+                    **self.sample_data,
+                    aliquots=self._create_aliquots(
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: This looks pretty messy and is hard to follow. Maybe try defining the aliquots first, then passing them in?

```
aliquots = self._create_aliquots()
Sample(**sample_data, aliquots=aliquots)
```

File: dataservice/util/data_gen/data_generator.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        # Create participants
+        self._create_participants(self.max_participants)
+        self.teardown()
+
+    def _create_participants(self, total):
+        """
+        creates participants with samples, demographics, and diagnoses
+        """
+        for i in range(total):
+            self.p = Participant(
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: p probably doesn't need to be a class property. Same comment as below about assigning samples and other relationships beforehand.

File: dataservice/util/data_gen/data_generator.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+                external_id='participant_{}'.format(i),
+                samples=self._create_samples(
+                    random.randint(
+                        self.min_samples,
+                        self.max_samples)),
+                demographic=self._create_demographics(i),
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: I think you want `_create_demographics(1)` since you can only have one demographic per participant.

File: dataservice/util/data_gen/data_generator.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+
+            self.e_data = {
+                'external_id': 'sequencing_experiment_{}'.format(i),
+                'experiment_date': dt - relativedelta.relativedelta(
+                    years=random.randint(1, 3)) + relativedelta.relativedelta(
+                    months=random.randint(1, 6)) +
+                relativedelta.relativedelta(days=random.randint(1, 30)),
+                'experiment_strategy':
+                random.choice
+                (self.experiment_strategy_list),
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: the function arguments should really be on the same line as the function

**Comment by dankolbman** [comment]
Should be sufficient for a first pass, but there's some stuff that could be cleaned up.
I'm not too big a fan of being heavily reliant on the class properties that are assigned within functions, eg:

```
class DataMocker:
    def __init__(self):
        self._genomic_files()

    def _genomic_files(self):
        self.max_gen_files = 10
        self.min_gen_files = 1
        self.file_format_list = ['.txt', '.csv']
```

You now have a bunch of class properties that follow different naming schemes all over the file, and you'd have to grep through to find the one you're looking for. You could perhaps create some sort of `MockEntity` class with `entity_class`, `min`, `max`, `sample_values` attributes for each entity, but that may be over abstracting in this case. You may be better off trying to use some sort of `dict` to do it instead:

```
entities = [ Participant, GenomicFile, ... ]
self.entity_ranges =  {c: {
    'min': 5,
    'max': 25,
    'choices': open(c.__class__.__name__+'.txt').read().split('\n')
} for c in entities }
```

Now you could do `self.entity_ranges[Participant]` and get back the min, max, and choices in a consistent manner.

Perhaps worth considering adding a `fake()` static function on each of the entities that returns a new entity with appropriate fields set.

[Here's](https://github.com/klen/mixer#flask-flask-sqlalchemy) a cool library that may give some inspiration, too.

File: dataservice/util/data_gen/data_generator.py (hunk 13 above)
Comment (parimalak) [reply dataservice/util/data_gen/data_generator.py]: i am just trying to have same suffix for creating demographic. 
external_id': 'demo_id_{}'.format(i)

File: dataservice/util/data_gen/data_generator.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+from dataservice.api.diagnosis.models import Diagnosis
+from dataservice.api.sample.models import Sample
+from dataservice.api.aliquot.models import Aliquot
+from dataservice.api.sequencing_experiment.models import SequencingExperiment
+from dataservice.api.genomic_file.models import GenomicFile
+
+
+class DataGenerator(object):
+
+    def __init__(self, config_name='testing'):
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: Can you set the config name to default to `FLASK_CONFIG` in the environment?

File: dataservice/util/data_gen/data_generator.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+from dataservice.api.demographic.models import Demographic
+from dataservice.api.diagnosis.models import Diagnosis
+from dataservice.api.sample.models import Sample
+from dataservice.api.aliquot.models import Aliquot
+from dataservice.api.sequencing_experiment.models import SequencingExperiment
+from dataservice.api.genomic_file.models import GenomicFile
+
+
+class DataGenerator(object):
+    def __init__(self, config_name=os.environ.get('FLASK_CONFIG', 'default')):
```
Comment (dankolbman) [inline dataservice/util/data_gen/data_generator.py]: Probably not a good idea to do any logic like this in a function handle. Probably more appropriate to put this in the `__init__.py` where the `DataGenerator` object get's created?

Comment (znatty22) [reply dataservice/util/data_gen/data_generator.py]: You could get rid of config_name as a param and assume it will be read in from the environ, something like: 
```python
class DataGenerator(object):
    def __init__(self):
        self.setup(os.environ.get('FLASK_CONFIG', 'default'))
```
OR leave the config_name as an optional parameter and later on it could be passed in as a cmd line arg to the flask cmd:
 
```python
class DataGenerator(object):
    def __init__(self, config_name=None):
        if not config_name:
            config_name = os.environ.get('FLASK_CONFIG','default')
        self.setup(config_name)
```

**Review by dankolbman — APPROVED** [review]
Nice!

**Review by znatty22 — APPROVED** [review]
